Live-live CI pipeline

trylek opened the pull request:
I have rebased the change against Santiago's CoreFX YAML refactoring
which simplified the change substantially. I have received an almost green
run and my cursory validation confirmed that live artifacts are indeed being
used.

I hit trouble on Linux x64 that doesn't seem to support Python3, I'm currently
experimenting with downgrading the patch-corefx.py script to not explicitly
target Python3. As a form of instrumentation to simplify development I'm
still hijacking the PR YAML file under CoreCLR for the "live-live pipeline", that's
not to be merged in, otherwise I believe the change to be mostly ready for
review.

I have slightly unified naming of the libraries jobs with the coreclr jobs. There's
a slight wrinkle in CoreCLR supporting three configurations (debug / checked
/ release) wherease CoreFX only supports debug / release. In this change
I have fixed CoreFX YAML scripts to also pretend to support the three configs,
just internally remapping "checked" to "debug" in base-job.yml.

[After all, in the live-live mode, there's technically a difference between debug
and checked version of CoreFX w.r.t. consuming the respective version of
CoreLib. I guess this logic is up to further development to support various
cross-building scenarios, e.g. debug CoreFX builds consuming release CoreLib.
I haven't tried to address these scenarios in my change but they should be
pretty straightforward to implement on top of it.]

As I assume that most people are on the Thanksgiving prolonged weekend,
I continue developing the change as I doubt I'll get it sufficiently reviewed
before next Monday. In particular, I have started working on incorporating
CoreFX testing. If any reviewers happen to be available (if I start receiving
review feedback), I'll fix the instrumentation by moving the YAML file and
I'll finalize the change so that it can be merged in.

Thanks

Tomas

trylek committed:
Live-live pipeline rebased against Santiago's CoreFX build change

trylek committed:
Fix library script name

trylek committed:
One more configuration mapping fix for CoreFX build

trylek committed:
Fix Windows build-test script, expand build matrix

trylek committed:
Add jobs for CoreFX test execution, experimentally downgrade Python

trylek commented:
I'm also not clear on what I've broken in the "coreclr-libraries" pipeline so if anyone can help shed light on that, I would be super grateful.

trylek committed:
Move download of live CoreCLR artifacts to base-job

trylek committed:
Don't clean up unpack folder in build-test-job as that nukes CoreCLR bits

trylek committed:
Fix path to sendtohelix.proj script

trylek commented:
Other than that, I must admit I find it almost impossible to conclusively prove that we're using the live-live artifacts everywhere with so much nuget downloading going on everywhere around; for now I tend to think this needs to be a collective effort after merging in the initial implementation of the live-live CI pipeline involving scrutiny by experts on build logic for the individual subrepos.

trylek commented:
I'm also trying to figure out how to incorporate core_setup aka installer testing in the pipeline but I know very little about what actually happens there - I tried to run the "installer" script locally and it seems to have Crossgened the framework and then crashed with some nuget download bug - so I'm in doubt whether I should follow up on that with help from the relevant experts or just merge in a limited version of the change without the installer chores and leave those to people more familiar with this logic.

trylek committed:
Specify libraries run more precisely to hopefully get some queues

trylek committed:
Instrumentation for chasing down missing Helix artifact (test runtime)

trylek committed:
Experimentally import Directory.Build.props from sendtohelix.proj

... to see whether it fixes the missing property definitions
for TestArchiveRuntimeRoot and others needed to compress the runtime
directory as part of Helix publishing.

trylek committed:
Bump up msbuild diagnostics for the sendtohelix script

trylek committed:
Fix configuration ID in calls to helix.yml

trylek committed:
Remove instrumentation; hopefully fix the coreclr-libraries pipeline

jaredpar commented:
> I hit trouble on Linux x64 that doesn't seem to support Python3, 

@jkoritzinsky where do we document our software requirements around tools like Python for Linux? The [Linux requirements](https://github.com/dotnet/runtime/blob/master/docs/workflow/linux-requirements.md) document doesn't specify this in the same way that the [Windows](https://github.com/dotnet/runtime/blob/master/docs/workflow/windows-requirements.md) one does.

trylek commented:
According to my best knowledge, this commit should end up with a green runtime-libraries and runtime-coreclr (at this moment still representing my initial implementation of the live-live pipeline). Remaining issues I'm aware of:

1) Where exactly to put the remapping layer between 3 CoreCLR configs (debug / checked / release) vs. 2 CoreFX configs (Debug / Release)? Today I'm doing this in several places in base-job, build-job, run-test-job and the CoreCLR build-test.sh and cmd scripts. I'm not sure if this is cleanly solvable right now but we should discuss the ideal longer-term state.

2) I had to open-code some of the variables in eng/pipelines/libraries/variables.yml (sourcesRoot, _buildScriptFileName) - as according to my discussion with Jared and Chris Costa from several months back there's no way to include the variable template "in the middle" of the template tree, just at the root (i.e. in "pipeline" scripts like pr.yml), there's no easy way to import it into the live-live pipeline spanning multiple subrepos.

3) I think I have managed to stand up the CoreFX test build and execution within the initial live-live pipeline implementation, for now I have abandoned the attempt to include the installer jobs in the pipeline as I'm simply not yet sufficiently familiar with them so that I would probably end up producing some unusable nonsense anyway.

4) Physically moving the live-live YAML file from its temporary location under eng/pipeline/coreclr/pr.yml (which is most convenient for developing the PR) to its final location; according to initial suggestion from Viktor is should probably become something like eng/pipelines/runtime.yml.

Thanks

Tomas

jaredpar commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -5,6 +5,7 @@ parameters:
   archType: ''
   framework: netcoreapp
   isOfficialBuild: false
+  useLiveCoreClr: false
```
Going forward when should this ever be false and what would the impact of that be?

trylek replied:
In this change I haven't yet dropped the support for both "live-live runs" and "partial runs" picking up some artifacts from external locations. I can easily imagine this limitation may be temporary but, in its current form, dropping this condition would invalidate the "legacy" runtime-libraries pipeline. I assume it's up to further discussion in our meetings how exactly to sequence the transformation.

jashook commented:
I have started to review. Based on my first pass this looks like great progress thank you @trylek. I have noticed that the coreclr pipeline is loosing coverage. I think this is fine, so long is it is documented and we can work on recovering this in later changes.

I will attempt to document what is being lost as part of my review. My suggestion as a min-bar for testing in this change would be:

1. linux checked arm64 build + test [checkbox:unchecked]
2. linux checked x64 build + test
3. OSX checked x64 build + test
4. windows checked x64 build + test
5. windows checked x86 build + test
6. windows checked arm build + test
7. windows checked arm64 build
8. windows release x64 build
9. linux release x64 build

jashook commented on eng/pipelines/coreclr/pr.yml:
```diff
@@ -4,24 +4,6 @@ pr:
     - master
-    - release/*.*
-  paths:
-    include:
```
I believe we will want to keep these exclude/include paths.

jashook commented on eng/pipelines/coreclr/pr.yml:
```diff
@@ -4,24 +4,6 @@ pr:
   branches:
     include:
     - master
-    - release/*.*
```
This should run on release/*.*

jashook reviewed (dismissed):
This is an overview feedback. This change looks like it is removing the idea of specific pipelines per area path. Either that or it is modifying `coreclr/pr.yml` into a root `pr.yml`. Either way I think we should change the direction to do the following:

1. Create a new root `pr.yml` without modifying `coreclr/py.yml`
   - The contents of this `pr.yml` should be similar to what is currently in this pr. I will make suggestions on what to remove assuming the pr.yml changed in this pr will be the new pr.yml for the root pipeline
2. Undo the changes to `coreclr/pr.yml` so that we will not lose coverage. If there is work to be done that is outside of the scope of what has been done already remove/comment out the bits that do not work such that we have minimum coverage detailed https://github.com/dotnet/runtime/pull/294#issuecomment-560473559
3. Installer will need to have work in its CI, which I expect to be less than corefx and coreclr. This should be done in a subsequent PR.

jashook commented:
I will continue a deeper review assuming that the `coreclr/pr.yml` will be essentially copied to become `root/pr.yml`

jashook commented on eng/pipelines/coreclr/pr.yml:
```diff
@@ -30,176 +12,111 @@ jobs:
 
 #
-# Checked JIT test executions
+# Test builds using live CoreFX
```
Past this point we can probably remove everything. The test builds should be done by `coreclr/pr.yml` and `libraries/pr.yml`

trylek replied:
As discussed in today meeting, my current assumption is that I'll rename this to a new pipeline (perhaps eng/pipelines/root.yml as I think you yourself suggested but I'll gladly listen to other suggestions) and for now I'll put back the original CoreCLR PR pipeline as is. With this modification, I think it makes sense to keep the CoreFX test logic in the new pipeline as it's something I've already fixed and can be reused (or rather kept and conditionally filtered out according to our plan from earlier today). Please let me know if you think otherwise.

jashook commented on src/coreclr/build-test.cmd:
```diff
@@ -577,8 +581,10 @@ REM ============================================================================
 
 if NOT "%__LocalCoreFXPath%"=="" (
     echo Patch CoreFX from %__LocalCoreFXPath%
+    set __CoreFXBuildType=Debug
```
This should be the opposite, assume it is release. If the build type is debug then use a debug corefx. That way we can correctly use a release corefx build with a checked runtime.

trylek replied:
Fixed, thanks for pointing that out. Ultimately we may want to introduce options for [semi-]arbitrarily combining configuration modes for the individual components (CoreCLR, CoreFX, tests in either subrepos - not sure what degrees of freedom we have) but I think it will be easier to address in a follow-up change.

jashook commented on src/coreclr/build-test.sh:
```diff
@@ -136,7 +136,12 @@ patch_corefx_libraries()
 
-    patchCoreFXArguments=("-clr_core_root" "${CORE_ROOT}" "-fx_root" "${__LocalCoreFXPath}" "-arch" "${__BuildArch}" "-build_type" "${__BuildType}")
+    __CoreFXBuildType=Debug
+    if [ "$__BuildType" == "release" ]; then
```
Same comment as the windows file.

trylek replied:
Fixed.

jashook reviewed (commented):
The rest looks good. The overview review sounds a bit daunting at first; however, the framework that you guys have built should make it I think pretty smooth to transition.

Thanks for all the great work.

jashook commented:
CI succeeded excellent work!

safern commented on eng/pipelines/libraries/base-job.yml:
```diff
@@ -14,23 +15,26 @@ parameters:
+        - buildScriptFileName: libraries
+        - _BuildConfig: Debug
+        - ${{ if eq(parameters.buildConfig, 'release') }}:
+          - _BuildConfig: Release
```
Should this be part of xplat-setup.yml?

safern replied:
Btw is comparison here case-sensitive? because in `libraries/.azure-ci.yml` we pass it down as upper-case.

safern commented on eng/pipelines/libraries/base-job.yml:
```diff
@@ -65,19 +69,30 @@ jobs:
         - ${{ if eq(parameters.osGroup, 'Windows_NT') }}:
           - _msbuildCommand: powershell -ExecutionPolicy ByPass -NoProfile eng\common\msbuild.ps1 -warnaserror:0 -ci
           - _runtimeOSArg: /p:RuntimeOS=win10
+          - _archiveExtension: '.zip'
```
Why do we need this? This should already be defined in `xplat-setup.yml`

trylek replied:
Ahh, you're right, this should be no longer necessary with all your refactorings, will remove, thank you!

safern commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -37,11 +44,13 @@ jobs:
         - ${{ if eq(parameters.runTests, true) }}:
           - _buildAction: -build -buildtests /p:ArchiveTests=true
-          - ${{ if ne(parameters.framework, 'allConfigurations') }}:
+          - ${{ if and(ne(parameters.framework, 'allConfigurations'), ne(parameters.useLiveCoreClr, true)) }}:
```
Why do we want to skip restoring the tests when `useLiveCoreClr` is not being used?

trylek replied:
I'll double-check this, I may have confused this with the other property to not restore the CoreCLR framework (as we're using the live one).

safern replied:
Yeah. That property is to avoid restoring libraries test-projects. AllConfigurations is a special framework which doesn't build libraries test projects at all (it builds custom package tests) so we don't need to restore test projects as part of its restore step.

safern commented on eng/pipelines/coreclr/pr.yml:
```diff
@@ -30,176 +12,111 @@ jobs:
-    helixQueueGroup: pr
-    helixQueuesTemplate: /eng/pipelines/coreclr/templates/helix-queues-setup.yml
+    - Windows_NT_x64
+    - Windows_NT_arm64
```
Tests aren't arch specific, so you should be able to just build `Windows_NT_x64`.

safern commented on eng/pipelines/coreclr/pr.yml:
```diff
@@ -30,176 +12,111 @@ jobs:
     - Windows_NT_x64
-    helixQueueGroup: pr
-    helixQueuesTemplate: /eng/pipelines/coreclr/templates/helix-queues-setup.yml
+    - Windows_NT_arm64
```
Tests aren't arch specific, so we should be able to just build them for x64 in all OSs.

dagood commented:
> I'm also trying to figure out how to incorporate core_setup aka installer testing in the pipeline but I know very little about what actually happens there - I tried to run the "installer" script locally and it seems to have Crossgened the framework and then crashed with some nuget download bug - so I'm in doubt whether I should follow up on that with help from the relevant experts or just merge in a limited version of the change without the installer chores and leave those to people more familiar with this logic.

I agree, I think it would be good if you can open an issue with the error you got so far (and note that it hasn't been merged yet so the error may be hard to reproduce until then). Since it apparently happens during restore I would probably have the best context on it to figure it out.

This also might be something that's already fixed in the "local live-live" PR.

trylek committed:
Addressing most of PR feedback

trylek commented:
I have created the new pipeline (runtime-live, feel free to rename it as you see fit):

https://dev.azure.com/dnceng/public/_build/results?buildId=444392

It cannot run green before my change is merged in.

trylek committed:
Fix unification of archiveExtension

jashook commented:
Changed the name to runtime :)

trylek commented:
Sure, my pleasure ;-)

jashook commented:
/cc @echesakovMSFT @sbomer

safern commented on eng/pipelines/root.yml:
```diff
@@ -0,0 +1,123 @@
+trigger: none
```
Should this instead be named `runtime.yml`?

safern commented on eng/pipelines/root.yml:
```diff
@@ -0,0 +1,123 @@
+      useLiveCoreClr: true
+
+#
+# CoreFX Test Execution
```
Please replace `CoreFX` with `Libraries`.

safern commented on eng/pipelines/root.yml:
```diff
@@ -0,0 +1,123 @@
+    - Linux_x64
+    - Windows_NT_x86
+    - Windows_NT_x64
+    - Windows_NT_arm64
```
Libraries tests, should just need to be built for x64, We can tweak the dependencies on run templates to depend on x64 always.

safern commented on eng/pipelines/libraries/base-job.yml:
```diff
@@ -65,19 +69,26 @@ jobs:
+
+        - ${{ if eq(parameters.useLiveCoreClr, true) }}:
+          - _coreClrDownloadPath: '$(Build.SourcesDirectory)/artifacts/transport/coreclr'
+          - _coreClrArtifactName: 'CoreCLRProduct_${{ parameters.osGroup }}${{ parameters.osSubgroup }}_${{ parameters.archType }}_${{ parameters.buildConfig }}'
```
Right now if I want to build a debug libraries build I would need to build coreclr for debug. Maybe we should accept a parameter which is `coreclrBuildConfig` in order to download the right artifact and depend on the right build.

safern commented on eng/pipelines/coreclr/templates/xplat-pipeline-job.yml:
```diff
@@ -103,6 +104,18 @@ jobs:
         - name: priorityArg
           value: '-priority=1'
 
+    - coreFxArtifactName: ''
```
coreFx -> libraries here as well. We might just want to do it for all the yml files.

safern commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -22,12 +23,18 @@ jobs:
+
+      ${{ if eq(parameters.useLiveCoreClr, true) }}:
+        dependsOn:
+        - ${{ format('coreclr_product_build_{0}{1}_{2}_{3}', parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig) }}
```
This would also need to allow depending on a different build config of coreclr.

jaredpar commented on eng/pipelines/libraries/base-job.yml:
```diff
@@ -14,23 +15,26 @@ parameters:
       pool: ${{ parameters.pool }}
       variables:
-        - _BuildConfig: ${{ parameters.buildConfig }}
+        - buildScriptFileName: libraries
```
Why do some of the variables here get prefixed with `_` and the others do not?

safern replied:
We should be prefixing this one as well. I believe he just moved it out from the variables template we have as it was. I usually prefix them with `_` since azure devops propagates the variables as environment variables and that can conflict with MSBuild and do weird stuff you didn't intend MSBuild to do because there was a variable name clash in between yaml variables and MSBuild properties.

trylek committed:
Santiago's feedback - rename coreFx to libraries in the new code

trylek committed:
Introduce liveCoreClrBuildConfig and liveLibrariesBuildConfig

These complementary properties should let us drop all automatic
transformations of CoreCLR vs. library build modes and thus
substantially simplify the live-live pipeline and make it more
general at the same time. I have reverted the changes to the
build-test scripts that were taking care of this logic on the
local build side as well as the change to
eng/pipelines/libraries/base-job and bits of the other YAML files
on the CI side.

Thanks

Tomas

trylek committed:
Rename root.yml to runtime.yml per PR feedback

trylek committed:
Add build-test argument to specify CoreFX configuration

trylek committed:
Remove msbuild verbosity instrumentation; fix YAML typo

jaredpar commented on eng/pipelines/runtime.yml:
```diff
@@ -0,0 +1,123 @@
+    - Linux_x64
+    - Windows_NT_x86
+    - Windows_NT_x64
+    - Windows_NT_arm64
```
The `platforms` section is identical between all the jobs here. Is there a reason we don't use a variable for this?

safern replied:
Since this is a parameter I believe this can get very messy on using a variable. Plus I believe when adding the full matrix and including libraries and installer full logic plus platform matrix in between CI and PRs, they will be somehow different based on what build we're running.

jaredpar commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -5,6 +5,7 @@ parameters:
   archType: ''
   framework: netcoreapp
   isOfficialBuild: false
+  liveCoreClrBuildConfig: ''
```
Does `liveCoreClrBuildConfig` represent the version of CoreCLR we should be building and using or simply using? Or stated differently does this control whether we build CoreCLR here or merely which already built CoreCLR we should be consuming?

safern replied:
The intention of this is to tell the `Libraries` build which CoreCLR build it should depend on. This way we can build `CoreCLR` for Release and then Libraries for Debug and depend on the CoreCLR release build, that way configurations are decoupled.

jaredpar replied:
Gotcha. @trylek consider adding a comment clarifying this.

safern commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -40,8 +47,10 @@ jobs:
+
       steps:
-        - script: $(_buildScript) -restore $(_buildArguments) $(_skipTestRestoreArg)
+        - script: $(_buildScript) -restore $(_buildArguments) $(_skipTestRestoreArg) $(_coreClrOverridePathArg)
```
I suggest the _coreClrOverridePathArg to be included in the `_buildArguments` directly as it needs to be passed down for the build step as well. The reason for that is the way currently libraries resolves coreclr. During the build, if we find a `ReferenceFromRuntime=System.Private.CoreLib` we will transform that to a `ProjectReference=runtime.depproj` which will restore coreclr's transport package by default if no CoreCLROVerridePath is passed down. So we need to pass it down to the build for the projects that reference corelib from runtime to restore the correct S.P.CoreLib.

This will change with @jkoritzinsky change as the goal is that coreclr just binplaces the assets directly where libraries expects them or to create a convention in between the two of them so that we no longer need CoreCLROverridePath.

trylek committed:
Address PR feedback

safern commented on eng/pipelines/coreclr/templates/build-test-job.yml:
```diff
@@ -75,6 +88,17 @@ jobs:
+          cleanUnpackFolder: false
+          artifactFileName: '$(librariesArtifactName)$(archiveExtension)'
+          artifactName: '$(librariesArtifactName)'
+          displayName: 'live-built CoreFX'
```
CoreFX -> Libraries.

safern commented on eng/pipelines/coreclr/templates/build-test-job.yml:
```diff
@@ -55,7 +64,11 @@ jobs:
+      dependsOn:
+      - ${{ format('coreclr_product_build_{0}{1}_{2}_{3}', parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig) }}
+      - ${{ if ne(parameters.liveLibrariesBuildConfig, '') }}:
+        - ${{ format('libraries_build_{0}_{1}{2}_{3}_{4}', 'netcoreapp', parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.liveLibrariesBuildConfig) }}
```
It might make sense to remove `netcoreapp` from the libraries build names when target framework is netcoreapp as libraries is the only subset that builds for other frameworks. That way we don't need to hard code it here. Something to consider on a follow up PR.

trylek committed:
Latest bits of PR feedback and disabling test runs for now

safern commented on eng/pipelines/runtime.yml:
```diff
@@ -0,0 +1,138 @@
+    - Windows_NT_arm64
+    jobParameters:
+      isOfficialBuild: false
+      testScope: empty
```
No need to pass this down here. The default is already `''` . This is just used when running the tests as part of the build-job.

trylek committed:
Stop passing testScope: empty to libraries build

safern commented on eng/pipelines/coreclr/templates/xplat-pipeline-job.yml:
```diff
@@ -103,6 +104,18 @@ jobs:
+    - librariesDownloadDir: ''
+
+    - ${{ if ne(parameters.liveLibrariesBuildConfig, '') }}:
+      - librariesArtifactName: ${{ format('libraries_bin_{0}_{1}{2}_{3}_{4}', 'netcoreapp', parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.liveLibrariesBuildConfig) }}
```
It might make sense to move these `artifactName` variables to xplat-setup template as it is used to upload and download and we already have all the required parameters at that point. Anyway, that can be done in a follow up PR, just calling it out so that we don't forget.

safern reviewed (approved):
LGTM from the yml changes. I don't have much context on the coreclr test scripts, but looks good to my small knowledge of it.

trylek committed:
Sigh - one last bit, propagation of condition wasn't working properly
